File: Tripfinger/ios/BookmarkService.m
#import "RCTBridgeModule.h"
#import "Reachability.h"
#import "AppDelegate.h"
@import Firebase;

@interface BookmarkService : NSObject <RCTBridgeModule>
@end

@implementation BookmarkService {
  BOOL initialized ;
  NSString * userId;
  NSString * listId;
}

static NSString * KEY_MWM_BOOKMARKS_MIGRATED = @"MWM_BOOKMARKS_MIGRATED";
static NSString * KEY_FIREBASE_USER_ID = @"FIR_USER_ID";
static NSString * DB_BOOKMARK_LISTS = @"bookmarkLists";
static NSString * DB_BOOKMARK_ITEMS = @"bookmarkItems";

+ (instancetype)sharedInstance
{
  static BookmarkService *sharedInstance = nil;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    sharedInstance = [[BookmarkService alloc] init];
  });
  return sharedInstance;
}

- (id)init
{
  self = [super init];
  initialized = NO;
  return self;
}

+ (void)delay:(int)seconds block:(void(^)())block
{
  dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (UInt64)(seconds * (double)(NSEC_PER_SEC))),
                 dispatch_get_main_queue(), block);

}

- (void)initializeFirebase
{
  if (initialized) {
    NSLog(@"Firebase already initialized.");
    return;
  }
  dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0ul);
  dispatch_async(queue, ^{
    NSLog(@"Initializing firebase");
    initialized = YES;
    [FIRApp configure];
    [FIRDatabase database].persistenceEnabled = YES;
    [self populateUserId];
    //  FIRDatabaseReference * ref = [db reference];
  });
}

- (void)populateUserId
{
  NSString * storeId = [[NSUserDefaults standardUserDefaults] stringForKey:KEY_FIREBASE_USER_ID];
  if (storeId != nil) {
    userId = storeId;
    [self populateListId];
  } else {
    [self createNewUserId];
  }
}

- (void)createNewUserId
{
  if ([Reachability isOnline]) {
    [[FIRAuth auth] signInAnonymouslyWithCompletion:^(FIRUser *_Nullable user, NSError *_Nullable error) {
      if (error != nil) {
        NSLog(@"req errored: %@", error);
        [BookmarkService delay:3 block:^{
          [self createNewUserId];
        }];
      } else {
        userId = user.uid;
        [[NSUserDefaults standardUserDefaults] setValue:user.uid forKey:KEY_FIREBASE_USER_ID];
        [self populateListId];
      }
    }];
  } else {
    [BookmarkService delay:3 block:^{
      [self createNewUserId];
    }];
  }
}

- (void)syncUserData {
  [[[[[FIRDatabase database] reference] child:DB_BOOKMARK_LISTS] child:userId] keepSynced:YES];
  [[[[[FIRDatabase database] reference] child:DB_BOOKMARK_ITEMS] child:listId] keepSynced:YES];
}

- (void)populateListId {
  FIRDatabaseReference * userListRef = [[[[FIRDatabase database] reference] child:DB_BOOKMARK_LISTS] child:userId];
  [userListRef observeSingleEventOfType:FIRDataEventTypeValue withBlock:^(FIRDataSnapshot * _Nonnull snapshot) {
    if (![snapshot exists]) {
      NSLog(@"No lists. Creating one.");
      [self createNewListId];
      return;
    }
    NSEnumerator * children = snapshot.children;
    for (FIRDataSnapshot * childSnapshot in children) {
      listId = childSnapshot.key;
      NSLog(@"Got list key: %@", childSnapshot.key);
    }
    [self createItemListener];
  }];
}

- (void)createNewListId
{
  NSDictionary * bookmarkList = @{@"name": @"My Places"};
  FIRDatabaseReference * listRef = [[[[[FIRDatabase database] reference] child:DB_BOOKMARK_LISTS] child:userId] childByAutoId];
  [listRef setValue:bookmarkList];
}

- (void)createItemListener
{
  [self syncUserData];
  FIRDatabaseReference * itemsRef = [[[[FIRDatabase database] reference] child:DB_BOOKMARK_ITEMS] child:listId];
  [itemsRef observeEventType:FIRDataEventTypeValue withBlock:^(FIRDataSnapshot * _Nonnull snapshot) {
    NSEnumerator * children = snapshot.children;
    NSMutableArray<NSDictionary*> * bookmarks = [[NSMutableArray alloc] init];
    for (FIRDataSnapshot * childSnapshot in children) {
      NSMutableDictionary * bookmarkDict = [[NSMutableDictionary alloc] initWithDictionary:childSnapshot.value];
      bookmarkDict[@"databaseKey"] = childSnapshot.key;
      [bookmarks addObject:bookmarkDict];
    }
    NSLog(@"Fetched %lu bookmarks.", bookmarks.count);
    [AppDelegate setBookmarks:bookmarks];
  }];  
}
  
- (NSString*)addBookmark:(NSDictionary*)bookmark {
  if (listId == nil) {
    return nil;
  }
  FIRDatabaseReference * itemRef = [[[[[FIRDatabase database] reference] child:DB_BOOKMARK_ITEMS] child:listId] childByAutoId];
  [itemRef setValue:bookmark];
  return [itemRef key];
}

- (void)removeBookmark:(NSString*)databaseKey {
  if (listId == nil) {
    return;
  }
  FIRDatabaseReference * itemRef = [[[[[FIRDatabase database] reference] child:DB_BOOKMARK_ITEMS] child:listId] child:databaseKey];
  [itemRef setValue:nil];
}

RCT_EXPORT_MODULE();

RCT_REMAP_METHOD(initializeFirebase, exportedInitializeFirebase)
{
  [BookmarkService.sharedInstance initializeFirebase];
}
  
RCT_REMAP_METHOD(addBookmark, addBookmark:(NSDictionary*)bookmark resolver:(RCTPromiseResolveBlock)resolve
                    rejecter:(RCTPromiseRejectBlock)reject) {
  NSString* key = [BookmarkService.sharedInstance addBookmark:bookmark];
  if (key == nil) {
    NSError* error = [NSError errorWithDomain:@"tripfinger.com" code:10 userInfo:nil];
    reject(@"no_events", @"There were no events", error);
  }
  resolve(key);
}
  
RCT_REMAP_METHOD(removeBookmark, exportedRemoveBookmark:(NSString*)databaseKey) {
  [BookmarkService.sharedInstance removeBookmark:databaseKey];
}

//RCT_EXPORT_METHOD(setNavBarHidden:(BOOL *)hidden)
//{
//  dispatch_async(dispatch_get_main_queue(), ^{
//    UIViewController * root = [UIApplication sharedApplication].delegate.window.rootViewController;
//    UINavigationController * nav = root.childViewControllers[0];
//    [nav setNavigationBarHidden:hidden animated:YES];
//  });
//}

@end

File: iphone/Maps/Classes/BookmarkService.swift
import Foundation
import FirebaseAuth
import FirebaseDatabase
import SwiftyJSON

protocol BookmarkListener: class {
  func bookmarksUpdated(bookmarks: [BookmarkItem])
}

class BookmarkService: NSObject {

  private static let KEY_MWM_BOOKMARKS_MIGRATED = "MWM_BOOKMARKS_MIGRATED"
  private static let KEY_FIREBASE_USER_ID = "FIR_USER_ID"
  private static let DB_BOOKMARK_LISTS = "bookmarkLists"
  private static let DB_BOOKMARK_ITEMS = "bookmarkItems"

  private var bookmarkedListings = Set<String>()
  private var userId: String?
  private var listId: String?
  private var delegate: BookmarkListener!
  
  init(delegate: BookmarkListener) {
    self.delegate = delegate
    super.init()
    FIRDatabase.database().persistenceEnabled = true
    populateUserId()
  }

  private func populateUserId() {
    let storedId = NSUserDefaults.standardUserDefaults().stringForKey(BookmarkService.KEY_FIREBASE_USER_ID)
    if let storedId = storedId {
      userId = storedId
      populateListId()
    } else {
      createNewUserId()
    }
  }
  
  private func createNewUserId() {
    if NetworkUtil.connectedToNetwork() {
      FIRAuth.auth()!.signInAnonymouslyWithCompletion { (user, error) in
        if error != nil {
          print("req errored: \(error)")
          SyncManager.delay(3, closure: self.createNewUserId)
        } else {
          self.userId = user!.uid
          NSUserDefaults.standardUserDefaults().setValue(user!.uid, forKey: BookmarkService.KEY_FIREBASE_USER_ID)
          self.populateListId()
        }
      }
    } else {
      SyncManager.delay(3, closure: createNewUserId)
    }
  }
  
  private func syncUserData() {
    FIRDatabase.database().reference()
      .child(BookmarkService.DB_BOOKMARK_LISTS)
      .child(userId!)
      .keepSynced(true)
    FIRDatabase.database().reference()
      .child(BookmarkService.DB_BOOKMARK_ITEMS)
      .child(listId!)
      .keepSynced(true)
  }
  
  private func populateListId() {
    let userListsRef = FIRDatabase.database().reference()
      .child(BookmarkService.DB_BOOKMARK_LISTS)
      .child(userId!)
    userListsRef.keepSynced(true)
    userListsRef.observeSingleEventOfType(.Value, withBlock: { (snapshot) -> Void in
      if !snapshot.exists() {
        print("no lists. creating one.")
        self.createNewListId()
        return
      }
      let children = snapshot.children
      while let childSnap: AnyObject = children.nextObject() {
        let childSnapshot = childSnap as! FIRDataSnapshot
        self.listId = childSnapshot.key
        print("Got list key: " + childSnapshot.key)
      }
      self.createItemListener()
    })
  }
  
  private func createNewListId() {
    let list = BookmarkList(name: "My Places")
    let listRef = FIRDatabase.database().reference()
      .child(BookmarkService.DB_BOOKMARK_LISTS)
      .child(userId!).childByAutoId()
    listRef.setValue(list.toDict())
  }
  
  private func createItemListener() {
    syncUserData()
    let itemsRef = FIRDatabase.database().reference()
      .child(BookmarkService.DB_BOOKMARK_ITEMS)
      .child(listId!)
    itemsRef.observeEventType(.Value, withBlock: { (snapshot) -> Void in
      let children = snapshot.children
      var bookmarks = [BookmarkItem]()
      var bookmarkedListings = Set<String>()
      while let childSnap: AnyObject = children.nextObject() {
        let childSnapshot = childSnap as! FIRDataSnapshot
        let bookmarkDict = childSnapshot.value as! [String : AnyObject]
        let bookmark = BookmarkItem(dict: bookmarkDict)
        bookmark.databaseKey = childSnapshot.key
        bookmarks.append(bookmark)
        if let listingId = bookmark.listingId {
          bookmarkedListings.insert(listingId)
        }
      }
      print("Fetched \(bookmarks.count) bookmarks.")
      self.bookmarkedListings = bookmarkedListings
      self.delegate.bookmarksUpdated(bookmarks)
    })
  }
  
  func isListingBookmarked(listingId: String) -> Bool {
    return bookmarkedListings.contains(listingId)
  }
  
  func addBookmark(bookmark: BookmarkItem) {
    guard let listId = listId else {
      return
    }
    let itemRef = FIRDatabase.database().reference()
      .child(BookmarkService.DB_BOOKMARK_ITEMS)
      .child(listId).childByAutoId()
    itemRef.setValue(bookmark.toDict())
  }
  
  func updateBookmark(bookmark: BookmarkItem) {
    guard let listId = listId else {
      return
    }
    guard let databaseKey = bookmark.databaseKey else {
      print("Tried to update bookmark without databaseKey set.")
      return
    }
    let itemRef = FIRDatabase.database().reference()
      .child(BookmarkService.DB_BOOKMARK_ITEMS)
      .child(listId).child(databaseKey)
    itemRef.setValue(bookmark.toDict())
  }

  func removeBookmarkForListing(listingId: String) {
    
  }

  func removeBookmark(bookmark: BookmarkItem) {
    guard let listId = listId else {
      return
    }
    guard let databaseKey = bookmark.databaseKey else {
      print("Tried to update bookmark without databaseKey set.")
      return
    }
    let itemRef = FIRDatabase.database().reference()
      .child(BookmarkService.DB_BOOKMARK_ITEMS)
      .child(listId).child(databaseKey)
    itemRef.setValue(nil)
  }
  
  func isBookmarkMigrationDone() -> Bool {
    return NSUserDefaults.standardUserDefaults().boolForKey(BookmarkService.KEY_MWM_BOOKMARKS_MIGRATED)
  }
  
  func setBookmarksMigrated() {
    NSUserDefaults.standardUserDefaults().setValue(true, forKey: BookmarkService.KEY_MWM_BOOKMARKS_MIGRATED)
  }
}